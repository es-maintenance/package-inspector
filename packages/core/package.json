{
  "name": "@package-inspector/core",
  "version": "0.0.0",
  "description": "Core functionality for @package-inspector",
  "main": "./dist/index.js",
  "types": "./dist/index",
  "license": "MIT",
  "scripts": {
    "build": "tsc -p .",
    "watch": "tsc -p . --watch",
    "test": "vitest"
  },
  "sideEffects": false,
  "dependencies": {
    "@npmcli/arborist": "^5.1.0",
    "@package-inspector/plugins": "*"
  },
  "devDependencies": {
    "@types/node": "17.0.27",
    "fixturify": "^2.1.1",
    "package-json-type": "^1.0.3",
    "tmp": "^0.2.1",
    "typescript": "4.6.3",
    "vitest": "^0.10.0"
  }
}
