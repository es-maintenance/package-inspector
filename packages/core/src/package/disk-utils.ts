import debug from 'debug';
import fs from 'fs';
import path from 'path';

const logger = debug('pi-core:disk-utils');

export function getAllFiles(
  dirPath: string,
  exclude?: RegExp,
  arrayOfFiles?: string[]
) {
  const files = fs.readdirSync(dirPath);

  const _arrayOfFiles = arrayOfFiles || [];

  files.forEach(function (file: string) {
    const fullPath = path.join(dirPath, file);

    try {
      if (
        fs.statSync(fullPath).isDirectory() &&
        (!exclude || (exclude && !exclude.test(fullPath)))
      ) {
        try {
          _arrayOfFiles.push(...getAllFiles(fullPath, exclude));
        } catch (ex: any) {
          logger(ex.message);
        }
      } else {
        if (!exclude || (exclude && !exclude.test(fullPath))) {
          _arrayOfFiles.push(fullPath);
        } else {
          logger('excluding', fullPath, ' for ', dirPath, exclude);
        }
      }
    } catch (ex: any) {
      logger(ex.message);
    }
  });

  return _arrayOfFiles;
}

export function getDirectorySize({
  directory,
  exclude,
}: {
  directory: string;
  exclude?: RegExp;
}): { files: number; physical: number } {
  const arrayOfFiles = getAllFiles(directory, exclude);

  const size = arrayOfFiles
    .map((filePath) => fs.statSync(filePath).size)
    .reduce((a, b) => a + b, 0);

  return {
    files: arrayOfFiles.length,
    physical: size,
  };
}

export function stripPathOnDisk(nodePath: string, workingPath: string): string {
  if (process.env.NODE_ENV === 'test') {
    // macOS uses a sym-link for `/tmp -> /private/tmp` dir and `nodePath` will start `/private`
    const slugToStrip = nodePath.startsWith('/private')
      ? `/private${workingPath}`
      : workingPath;

    return nodePath.replace(slugToStrip, '');
  }

  return nodePath;
}
