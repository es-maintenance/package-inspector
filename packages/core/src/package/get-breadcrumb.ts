import { IArboristNode } from '../types';

export function getBreadcrumb(node: IArboristNode): string {
  const bread: string[] = [];

  function walk(node: IArboristNode): string[] {
    if (bread.includes(node.name)) {
      return bread;
    }

    if (node.edgesIn) {
      const [edge] = [...node.edgesIn.values()];

      if (edge && edge.from) {
        bread.push(edge.name);

        return walk(edge.from);
      } else {
        // we have gotten to the root project, don't push the root project name
        return bread;
      }
    } else {
      return bread;
    }
  }

  return walk(node).reverse().join('#');
}
