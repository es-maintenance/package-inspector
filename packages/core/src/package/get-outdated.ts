import { exec } from 'child_process';
import { promisify } from 'util';

const execPromise = promisify(exec);

export interface OutdatedData {
  [key: string]: {
    latest: string;
  };
}

export async function getOutdated(dir: string): Promise<OutdatedData> {
  let packageData: OutdatedData = {};

  try {
    await execPromise('npm outdated --json', {
      cwd: dir,
    });
  } catch (ex: any) {
    packageData = JSON.parse(ex.stdout.toString('utf8')) as OutdatedData;
  }

  return packageData;
}
