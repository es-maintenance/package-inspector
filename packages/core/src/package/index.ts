export { getAllFiles, getDirectorySize } from './disk-utils';
export { getBreadcrumb } from './get-breadcrumb';
export { getLatestPackages } from './get-latest-packages';
export { humanFileSize } from './human-file-size';
