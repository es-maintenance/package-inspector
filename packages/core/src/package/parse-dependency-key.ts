// @babel/core@0.5.0
// core-js@0.5.0
export function parseDependencyKey(key: string): {
  name: string;
  version?: string;
} {
  const parts = key.split('@');

  // We found an `@` symbol at the beginning of the key, so we have a scoped package name without a version.
  if (parts.length === 2 && parts[0] === '') {
    return {
      name: key,
    };
  }

  // We hit a case like `@babel/core@0.5.0`
  if (parts.length === 3) {
    return {
      name: `${parts[0]}@${parts[1]}`,
      version: parts[2],
    };
  }

  return {
    name: `${parts[0]}`,
    version: parts[1],
  };
}
