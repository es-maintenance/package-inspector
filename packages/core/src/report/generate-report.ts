import Arborist from '@npmcli/arborist';
import path from 'path';

import { Report } from '../models';
import {
  formatDuration,
  getDirectorySize,
  getLatestPackages,
  stripPathOnDisk,
} from '../package';
import { IArboristNode, ServerPlugin } from '../types';

function getValues(dependencyTree: IArboristNode) {
  // ignore the root node
  return [...dependencyTree.inventory.values()].filter(
    (node) => !node.isProjectRoot
  );
}

function processPlugins(plugins: string[]): ServerPlugin[] {
  const processedPlugins: ServerPlugin[] = [];

  plugins.forEach((pluginPath) => {
    // FIXME: need to have a try catch for this
    // FIXME: we need to have a test for this code, easily breakable
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const ServerPlugin = require(`${pluginPath}/server`)?.default;

    if (ServerPlugin) {
      processedPlugins.push(new ServerPlugin());
    }
  });

  return processedPlugins;
}

interface GenerateReportArgs {
  plugins?: string[];
}

export async function generateReport(
  cwd: string,
  options?: GenerateReportArgs
): Promise<Report> {
  let processedPlugins: ServerPlugin[] = [];

  if (options?.plugins) {
    processedPlugins = processPlugins(options.plugins);
  }

  const report = new Report();

  const arb = new Arborist({
    path: cwd,
  });

  const rootArboristNode: IArboristNode = await arb.loadActual();
  const arboristValues = getValues(rootArboristNode);

  const latestPackagesMap = await getLatestPackages(arboristValues);

  Object.keys(latestPackagesMap).forEach((packageName) => {
    report.latestPackages[packageName] = latestPackagesMap[packageName];
  });

  arboristValues.forEach((depNode) => {
    const edge = [...depNode.edgesIn.values()].find((edgeIn) => {
      return edgeIn.name === depNode.name;
    });

    const lookupKey = `${depNode.name}@${depNode.version}`;

    if (!report.dependencies[lookupKey]) {
      report.dependencies[lookupKey] = {
        funding: depNode.funding,
        homepage: depNode.homepage,
        name: depNode.name,
        version: depNode.version,
        type: edge?.type || (depNode.isWorkspace ? 'workspace' : undefined),
        metadata: {
          size: getDirectorySize({
            directory: depNode.path,
            exclude: new RegExp(path.resolve(depNode.path, 'node_modules')),
          }),
          pathsOnDisk: [],
        },
        dependencies: [...depNode.edgesOut.values()]
          .filter((dependency) => {
            const node = dependency.to;

            // We don't have a node when it is optional or a peer
            // At this point we don't count those as they are counted as a part of the package that brings them in
            return node;
          })
          .map((dependency) => {
            const node = dependency.to;

            return `${node.name}@${node.version}`;
          }),
      };
    }

    // always push the path of the depNode, everything is taken care of by the setup code.
    report.dependencies[lookupKey].metadata?.pathsOnDisk.push(
      stripPathOnDisk(depNode.path, cwd)
    );
  });

  report.root = {
    name: rootArboristNode.packageName,
    version: rootArboristNode.version,
    funding: rootArboristNode.funding,
    homepage: rootArboristNode.homepage,
    metadata: {
      pathsOnDisk: [stripPathOnDisk(rootArboristNode.path, cwd)],
    },
    dependencies: [...rootArboristNode.edgesOut.values()]
      .filter((dependencyEdge) => {
        const dependencyNode = dependencyEdge.to;

        // We don't have a node when it is optional or a peer
        // At this point we don't count those as they are counted as a part of the package that brings them in
        return dependencyNode;
      })
      .map((dependencyEdge) => {
        const dependencyNode = dependencyEdge.to;

        return `${dependencyNode.name}@${dependencyNode.version}`;
      }),
  };

  const suggestionInput = {
    arboristValues,
    rootArboristNode,
    dependencies: report.dependencies,
  };

  for (const plugin of processedPlugins) {
    if (plugin.getTasks) {
      const suggestionTasks = plugin.getTasks();

      for (const suggestionTask of suggestionTasks) {
        const suggestionOutput = await suggestionTask.run(suggestionInput);

        if (suggestionOutput) {
          report.suggestions.push(suggestionOutput);
        }
      }

      console.log();
      console.log(`Timing for tasks for: ${plugin.name}`);
      for (const suggestionTask of suggestionTasks) {
        console.log(
          `${suggestionTask.name} (${formatDuration(
            suggestionTask.elapsedTime
          )})`
        );
      }
      console.log();
    }
  }

  return report;
}
