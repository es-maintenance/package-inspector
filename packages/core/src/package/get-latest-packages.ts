import debug from 'debug';
import { copyFile, writeFile } from 'fs/promises';
import { IDependencyMap } from 'package-json-type';
import path from 'path';
import tmp from 'tmp';

import { IArboristNode } from '../types';
import { getOutdated } from './get-outdated';

let _CACHEDLATESTPACKAGES: IDependencyMap = {};

export async function getLatestPackages(
  arboristValues: IArboristNode[]
): Promise<IDependencyMap> {
  // only do this once
  if (Object.keys(_CACHEDLATESTPACKAGES).length) {
    return _CACHEDLATESTPACKAGES;
  }
  // TODO: this should be cached for faster build times and not having to make large requests to NPM
  const fakePackageJson: { dependencies: IDependencyMap } = {
    dependencies: {},
  };

  const dependencyKeys = arboristValues
    .filter((node) => {
      // ignore linked packages and packages that have realpaths that are on disk (which means they are linked and potentially don't exist in the registry)
      return (
        (!node.isLink || !node.isWorkspace) &&
        node.realpath.includes('node_modules')
      );
    })
    .map((node) => node.name);

  dependencyKeys.forEach((key) => {
    if (key.includes('fastlane')) return;
    fakePackageJson.dependencies[key] = '*';
  });

  const tmpobj = tmp.dirSync({ unsafeCleanup: true });
  debug('Dir: ' + tmpobj.name);

  await writeFile(
    path.resolve(tmpobj.name, 'package.json'),
    JSON.stringify(fakePackageJson)
  );

  const latestPackages: IDependencyMap = {};
  try {
    // need to keep the registry context for the fake package.json so things get resolved correctly when checking outdated
    await copyFile(
      path.resolve(process.cwd(), '.npmrc'),
      path.resolve(tmpobj.name, '.npmrc')
    );
  } catch (ex) {
    debug('Could not copy .npmrc, using the default npm registry.');
  }

  const outdatedPackages = await getOutdated(tmpobj.name);

  for (const packageName in outdatedPackages) {
    latestPackages[packageName] = outdatedPackages[packageName].latest;
  }

  tmpobj.removeCallback();
  _CACHEDLATESTPACKAGES = latestPackages;

  return latestPackages;
}
