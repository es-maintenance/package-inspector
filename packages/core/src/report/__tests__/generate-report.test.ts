import {
  afterEach,
  beforeAll,
  describe,
  expect,
  SpyInstance,
  test,
  vi,
} from 'vitest';
import fixturify, { type DirJSON } from 'fixturify';
import tmp from 'tmp';

import NO_NODE_MODULES from '../../__tests__/__fixtures__/no-node-modules.json';
import MIN_NODE_MODULES from '../../__tests__/__fixtures__/min-node-modules.json';
import { generateReport } from '../generate-report';
import * as GetOutdatedModule from '../../package/get-outdated';

describe('generate-report', () => {
  let getOutdatedSpy: SpyInstance;

  beforeAll(() => {
    tmp.setGracefulCleanup();
  });

  afterEach(() => {
    getOutdatedSpy.mockRestore();
  });

  test('it generates a report for a package with no node_modules', async () => {
    const { name: tmpLocation } = tmp.dirSync();

    fixturify.writeSync(tmpLocation, NO_NODE_MODULES as DirJSON);

    getOutdatedSpy = vi
      .spyOn(GetOutdatedModule, 'getOutdated')
      .mockImplementation(async () => ({}));

    const r = await generateReport(tmpLocation);

    expect(r).toMatchSnapshot();
  }, 10000);

  test('it generates a report for a package with basic node_modules', async () => {
    const { name: tmpLocation } = tmp.dirSync();

    fixturify.writeSync(tmpLocation, MIN_NODE_MODULES as DirJSON);

    getOutdatedSpy = vi
      .spyOn(GetOutdatedModule, 'getOutdated')
      .mockImplementation(async () => ({}));

    const r = await generateReport(tmpLocation);

    expect(r).toMatchSnapshot();
  }, 10000);

  test('it should add latest package info if outdated packages are found', async () => {
    const { name: tmpLocation } = tmp.dirSync();

    fixturify.writeSync(tmpLocation, NO_NODE_MODULES as DirJSON);

    // Setting up `getOutdated` to return a value
    getOutdatedSpy = vi
      .spyOn(GetOutdatedModule, 'getOutdated')
      .mockImplementation(async () => {
        return {
          foo: {
            latest: '0.2.1',
          },
        };
      });

    const r = await generateReport(tmpLocation);

    expect(r).toMatchSnapshot();
  });
});
