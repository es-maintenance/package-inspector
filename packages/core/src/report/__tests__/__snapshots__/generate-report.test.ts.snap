// Vitest Snapshot v1

exports[`generate-report > it generates a report for a package with basic node_modules 1`] = `
Report {
  "dependencies": {
    "left-pad@1.3.0": {
      "dependencies": [],
      "funding": "N/A",
      "homepage": "N/A",
      "metadata": {
        "pathsOnDisk": [
          "/node_modules/left-pad",
        ],
        "size": {
          "files": 10,
          "physical": 9752,
        },
      },
      "name": "left-pad",
      "type": "prod",
      "version": "1.3.0",
    },
  },
  "latestPackages": {},
  "pluginInfo": {},
  "root": {
    "breadcrumb": "",
    "dependencies": [],
    "funding": "N/A",
    "homepage": "N/A",
    "metadata": {
      "pathsOnDisk": [
        "",
      ],
    },
    "name": "min-node-modules",
    "version": "1.0.0",
  },
  "suggestions": [
    {
      "actions": [],
      "id": "packagesWithPinnedVersions",
      "message": "There are currently 0 pinned dependencies",
      "name": "Packages with pinned dependencies",
    },
    {
      "actions": [],
      "id": "packagesWithExtraArtifacts",
      "message": "There are currently 0 extra artifacts",
      "name": "Packages with extra artifacts",
    },
    {
      "actions": [],
      "id": "notBeingAbsorbedByTopLevel",
      "message": "There are currently 0 duplicate packages being installed on disk",
      "name": "Dependencies not being absorbed",
    },
    {
      "actions": [],
      "id": "nestedDependencyFreshness",
      "message": "Out of the total 1 sub packages currently installed; 0 major versions out of date (0.00%), 0 minor versions out of date (0.00%), 0 patch versions out of date (0.00%)",
      "name": "Nested Dependency Freshness",
    },
    {
      "actions": [],
      "id": "topLevelDepsFreshness",
      "message": "Out of the total 1 explicit dependencies defined in the package.json; 0 major versions out of date (0.00%), 0 minor versions out of date (0.00%), 0 patch versions out of date (0.00%)",
      "name": "Top Level Dependency Freshness",
    },
  ],
}
`;
